src/core/address_payload_unification_design.md

```diff
@@ -0,0 +1,556 @@
+  - It also seems clear that cases (3) and (4) should have the same
+    semantics as each other, but it is less clear what these should be.
+  - In these cases, we could elect to either return `undefined` or to
+    return a “null object”: an object for which `get()` returns `null`
+    and `neighbors(_)` returns `[].values()`.
+  - We should also ask—is the address required to be associated with a
+    known plugin? If so, should we ask the plugin to give us a
+    `NodeReference` instance specific to that plugin?
+  - If we later learn about this node, can the formerly dangling
+    reference come “back online”? Should it?
+  - Does this mean that calling `nodeReference` can change the internal
+    state of the graph? How does this impact the graph’s ability to
+    compress its `_nodes`?
+
+Some of these latter questions apply in case (2), too.
+
+One resolution to, I think, all of these questions is to change the
+`PluginHandler.createReference` method so that it does not accept a
+`NodePayload`, or at least to make the payload optional. This has the
+nice property that `PluginHandler<NR, NP>` has two simple functions: one
+to enrich a base reference into an `NR`, and one to enrich a JSON blob
+into an `NP`. I suspect that this is likely a good solution, but I am
+interested to hear other opinions.
+
+## Conclusion and next steps
+
+This API encompasses an improvement to how nodes are handled. Clients
+will usually work with either a `Node<NR, NP>` or a `NodeReference`
+alone. There is still a bit of friction in that some methods are
+inherently on the reference while others are on the porcelain, but this
+is not actually a problem when the `Node<NR, NP>` is treated as a
+package (not destructured): consider
+
+```javascript
+function someActions(
+  someRef: PullRequestReference,
+  someNode: Node<PullRequestReference, PullRequestPayload>
+) {
+  someRef.comments();        // easy
+  someNode.ref.comments();   // no problem
+  someNode.payload.title();  // piece of cake
+}
+```
+
+You have to pay a few extra characters for `.ref` or `.payload`, but the
+typechecker will help you out.
+
+We may want to do something similar for edges in the future, but also
+perhaps not. Edges are definitely different from nodes: we never really
+reference them by address (`Graph.edge` is currently never called
+outside of test code and `graph.js` itself), so something more like
+content-addressing might make more sense. But that is not an entirely
+straightforward change from our current model.
+
+I (@wchargin) still think that we should plan to eventually remove
+`type` from being privileged on the address. The `type` and `id`
+properties together actually represent a plugin-specific set of values
+that are used to uniquely determine the identity of a node, and which we
+therefore require to always be available—even when the full contents of
+a node are not loaded. This grants us more flexibility: perhaps it makes
+more sense to have
```

Comment:
In the examples below, you make a decent case for allowing other properties to be always included in the address-layer for a node (ie. required to get a reference). However, none of these examples make a case for removing type from its privileged position. Without type, it would be difficult to use those extra properties - if you can't filter by type, how do you know which properties will be available to you? Duck typing?

Our `compat` framework encodes the idea of type, and I think for very good reason. I think we should consistently enshrine type across the system, where type is not a string but a [pluginName: string, subtype: string] 2-tuple, or perhaps a [pluginName, subtypes....] k-tuple. It's actually the `id` that should be de-privileged. Git plugin types should have the option of using the `hash` as the id, or in the case of e.g. TreeEntrys using the union of `treeHash` and `entryName`. GitHub author-like nodes should be able to use `login` as id. And so forth.

Reply:
> However, none of these examples make a case for removing type from its
> privileged position. Without type, it would be difficult to use those
> extra properties - if you can't filter by type, how do you know which
> properties will be available to you? Duck typing?

I don’t follow the logic here (and I’m certainly not advocating for duck
typing). Who says “you can’t filter by type”? You’d be allowed to filter
by `type` just like any other property. If your filter includes a `type`
constraint and a node doesn’t have a `type` property, then the node is
not included in the results.

To put it another way: in a world where plugins can make their own
choices about what information is used to identify a node, there’s no
longer any _need_ to privilege `type`. I’m not trying to knock `type`
down a peg; I’m trying to raise everything else up. :-)